foreshadow/cleaners/data_cleaner.py

```diff
@@ -1,26 +1,279 @@
+            not possible, it will be None.
+            .match_lens
+            the number of characters from original text at each step that
+            was transformed.
 
         """
-        super().__init__()
+        matched_lengths = []  # this does not play nice with creating new
+        # columns
+        for transform in self.transformations:
+            row = row_of_feature
+            row, match_len = transform(row, return_search=True)
+            if match_len == 0:
+                matched_lengths.append(0)
+                row = self.default(row_of_feature)
+                break
+            matched_lengths.append(match_len)
+        if return_tuple:
+            return CleanerReturn(row, matched_lengths)
+        else:
+            return row
 
-    def fit(self, X, y=None, **fit_params):
-        """Stub fit method.
+    def fit(self, X, y=None):
+        """Empty fit.
 
         Args:
-            X: input data
-            y: labels
-            **fit_params: params to fit method.
+            X: input observations
+            y: input labels
+
+        Returns:
+            self
 
         """
-        pass
+        return self
+
+    def transform(self, X, y=None):
+        """Clean string columns.
+
+        Here, we assume that any list output means that these are desired
+        to be new columns in our dataset. Contractually, this could change
+        to be that a boolean flag is passed to indicate when this is
+        desired, as of right now, there should be no need to return a list
+        for any case other than this case of desiring new column.
+
+        The same is assumed for dicts, where the key is the new column name,
+        the value is the value for that row in that column. NaNs
+        are automatically put into the columns that don't exist for given rows.
+
+        Args:
+            X (:obj:`pandas.Series`): X data
+            y: input labels
+
+        Returns:
+            :obj:`pandas.DataFrame`: Transformed data
+
+        """
+        X = check_df(X, single_column=True)
+        # Problem:
+        # I can use .apply to perform all these transformations and that
+        # works beautifully, except when I want to define a funtion that
+        # will use the pandas.series.str.split operation. In which case,
+        # the .apply fails and I don't know why.
+
+        # I need each function to accept the row as an argument so that we
+        # can inspect how much of the text was matched (for determining if
+        # it should be used). however, doing this means I need to iterate
+        # over each row for a given column on my own, which requires me to
+        # leave
+
+        out = X[X.columns[0]].apply(self, return_tuple=False)  # access single
+        # column as series and apply the list of transformations to each row
+        # in the series.
+        if any(
+            [
+                isinstance(out[i], (list, tuple))
```

Comment:
out[i] isn't guaranteed to get the correct row, maybe you were looking for iloc?